integrationtests/gitjob/git/git_test.go

```diff
@@ -60,311 +64,343 @@ var (
+			// We attempt cleanup but don't fail the test if we can't delete due to permission issues.
+			// The test container framework will eventually clean up the temp directory.
+			_ = os.RemoveAll(tmpDir)
+		}
+	})
 
-	return func(code int) {
-		terminateContainer(ctx, t, container, code)
-	}
-}
+	DeferCleanup(func() {
+		if container != nil {
+			cleanCtx := context.Background()
+
+			// Gather logs for troubleshooting if test failed
+			if CurrentSpecReport().Failed() {
+				logs, err := container.Logs(cleanCtx)
+				if err == nil {
+					logContent, err := io.ReadAll(logs)
+					if err == nil {
+						GinkgoWriter.Printf("\n=== Container logs for Gogs ===\n%s\n", string(logContent))
+					}
+					_ = logs.Close()
+				}
+			}
 
-func TestLatestCommit_NoAuth(t *testing.T) {
-	ctlr := gomock.NewController(t)
-	defer ctlr.Finish()
-	ctx := context.Background()
+			err := container.Terminate(cleanCtx)
+			Expect(err).NotTo(HaveOccurred())
+		}
+	})
+})
 
-	tests := map[string]struct {
-		gitrepo        *v1alpha1.GitRepo
-		expectedCommit string
-		expectedErr    error
-	}{
-		"public repo": {
-			gitrepo: &v1alpha1.GitRepo{
-				Spec: v1alpha1.GitRepoSpec{
-					Repo:   url + "/test/public-repo",
-					Branch: "master",
-				},
-			},
-			expectedCommit: latestCommitPublicRepo,
-			expectedErr:    nil,
-		},
-		"private repo": {
-			gitrepo: &v1alpha1.GitRepo{
-				Spec: v1alpha1.GitRepoSpec{
-					Repo:   url + "/test/private-repo",
-					Branch: "master",
-				},
-			},
-			expectedCommit: "",
-			expectedErr:    transport.ErrAuthenticationRequired,
-		},
-	}
+var _ = Describe("Git Fetch", func() {
+	Describe("LatestCommit NoAuth", func() {
+		var ctx context.Context
+		var ctlr *gomock.Controller
 
-	config.Set(&config.Config{})
-	for name, test := range tests {
-		t.Run(name, func(t *testing.T) {
-			f := git.Fetch{}
-			client := mocks.NewMockK8sClient(ctlr)
-			// May be called multiple times, including calls to get Rancher CA bundle secrets
-			client.EXPECT().Get(ctx, gomock.Any(), gomock.Any()).
-				Return(apierrors.NewNotFound(schema.GroupResource{}, "notfound")).AnyTimes()
-			latestCommit, err := f.LatestCommit(ctx, test.gitrepo, client)
-			if test.expectedErr == nil {
-				require.NoError(t, err)
-			} else {
-				require.Contains(t, err.Error(), test.expectedErr.Error())
-			}
-			if latestCommit != test.expectedCommit {
-				t.Errorf("latestCommit doesn't match. got %s, expected %s", latestCommit, test.expectedCommit)
-			}
+		BeforeEach(func() {
+			ctx = context.Background()
+			ctlr = gomock.NewController(GinkgoT())
+			config.Set(&config.Config{})
 		})
-	}
 
-}
+		AfterEach(func() {
+			ctlr.Finish()
+		})
 
-func TestLatestCommit_BasicAuth(t *testing.T) {
-	ctlr := gomock.NewController(t)
-	defer ctlr.Finish()
-	ctx := context.Background()
+		type testCase struct {
+			repoPath          string
+			expectError       bool
+			getExpectedCommit func() string
```

Comment:
nit: why do we need a function here, as opposed to a simple `string` value (which may be empty)?

Reply:
I see, thanks for clarifying :bow: